### ISSUE TITLE
Questions about inconsistency between paper and code

### ISSUE THREAD

**Issue by BabelTower** [zhiming-xu/conad#2]
I had a problem when I reproduced your paper. 

In your paper, backbone consists of two layers of GAT, while in the open source code on Github, backbone consists of two layers of GCN and one layer of MLP.

I want to know the reason for the difference. Will the latter perform any better?

**Comment by zhiming-xu** [comment]
Hi Zeyu, thank you for noting the difference. This repo was working-in-progress code. So, it was not exactly the same as the paper. But in general, I think using either GCN or GAT in anomaly detection will yield similar performance.

**Comment by BabelTower** [comment]
```
def train_siamese(dataset, cuda=True, epoch1=100, epoch2=50, lr=1e-3, margin=0.5):
       ...
       criterion = lambda z, z_hat, l: torch.square(z - z_hat) * (l==0) - l * torch.square(z - z_hat) + margin
       ...
```
<img width="728" alt="image" src="https://user-images.githubusercontent.com/44083490/178295721-633213fd-e20a-4b1f-9d51-614b03c22907.png">

Is there a problem with the Siamese loss function?

Here the max function is missing, and 'margin' as a constant added to the loss function is meaningless.

**Comment by zhiming-xu** [comment]
Thank you for noting this. In later implementation, I changed the lambda function to torch.nn.MarginRankingLoss. I think you can check a recent implementation of the model in [pygod](https://github.com/pygod-team/pygod/blob/main/pygod/models/conad.py). It's written by me and should be more refined. I'll also try to fix the code here.

**Comment by BabelTower** [comment]
I saw this implementation earlier too, but I didn't quite understand this part.

```
margin_loss = self.margin_loss_func(h, h, h_aug) * label_aug
margin_loss = torch.mean(margin_loss)
```

The PyTorch documentation explains this criterion as follows:

$loss(x1,x2,y)=max(0,−y∗(x1−x2)+margin)$

Corresponding to the code, x1 is h, x2 is h, and y is h_aug, so the part $−y∗(x1−x2)$ is always going to be zero.

Hope you can point it out if I misunderstood. Thanks for your previous answer.

**Comment by zhiming-xu** [comment]
I think I found the problem here. In pygod implementation, I intended to use TripletMarginLoss but mistook it as MarginRankingLoss. A name change might fix the problem there. TripletMarginLoss can also accommodate the triplet loss in the paper while it works as a Siamese loss by taking `h` for the first two parameters. I do remember that I spent some time switching between the two loss functions. Thank you very much for finding out the problem. You can also open PR here and/or to pygod to add your fix.

> ```
> def train_siamese(dataset, cuda=True, epoch1=100, epoch2=50, lr=1e-3, margin=0.5):
>        ...
>        criterion = lambda z, z_hat, l: torch.square(z - z_hat) * (l==0) - l * torch.square(z - z_hat) + margin
>        ...
> ```
> 
> <img alt="image" width="728" src="https://user-images.githubusercontent.com/44083490/178295721-633213fd-e20a-4b1f-9d51-614b03c22907.png">
> 
> Is there a problem with the Siamese loss function?
> 
> Here the max function is missing, and 'margin' as a constant added to the loss function is meaningless.

**Comment by BabelTower** [comment]
I'm having some difficulty reproducing your paper, and when I implement the code exactly as it says in the paper, it's always difficult to achieve the desired accuracy.

Could you please provide me with a code corresponding to the paper? If possible, please send it to my email 2455255381@qq.com. Thank you very much.

**Comment by BabelTower** [comment]
I found that the problem is caused by the **structure decoder**. 

If the **sigmoid** activation function is used as mentioned in the experimental setting of the paper, the performance of the model is not very good. 

If I remove this activation function like in **pygod**, the model will perform much better.

I don't understand the reason for this. Why did you remove the activation function in pygod?

**Comment by zhiming-xu** [comment]
I'm sorry for the late response. I think the structure decoder is much more difficult to train than the attribute one. I believe it's because the adjacency matrix is extremely sparse with almost all entries being zeros. Therefore, a specific choice of activation function (and how their gradient looks like) can significantly impact the resulting performance.
